crates/matrix-sdk-ui/src/timeline/mod.rs

```diff
@@ -605,13 +622,21 @@ impl Timeline {
     /// reply also belongs to the unthreaded timeline. No threaded receipt
     /// will be sent here (see also #3123).
     ///
-    /// Returns a boolean indicating if we sent the request or not.
+    /// This also unsets the unread marker of the room if necessary.
+    ///
+    /// Returns a boolean indicating if it sent the receipt or not.
     #[instrument(skip(self), fields(room_id = ?self.room().room_id()))]
     pub async fn mark_as_read(&self, receipt_type: ReceiptType) -> Result<bool> {
         if let Some(event_id) = self.controller.latest_event_id().await {
             self.send_single_receipt(receipt_type, ReceiptThread::Unthreaded, event_id).await
         } else {
             trace!("can't mark room as read because there's no latest event id");
+
+            if self.room().is_marked_unread() {
+                // Unset the read marker.
+                self.room().set_unread_flag(false).await?;
```

Comment:
Can we add `let room = self.room();` to avoid calling `self.room()` multiple time please?

Reply:
With the removal of the condition, there is only one call here now.